### ISSUE TITLE
Does not deduce the type of a wrapped function assigned to a class variable correctly

### ISSUE THREAD

**Issue by kovidgoyal** [astral-sh/ty#3961]
### Summary

Here is a Python snippet that type checks cleanly on mypy, works correctly at runtime and fails on ty:
```py
from collections.abc import Callable

def wrapper[**P, R](method: Callable[P, R]) -> Callable[P, R]:
    def inner(*args: P.args, **kwargs: P.kwargs) -> R:
        return method(*args, **kwargs)
    return inner

class Class:
    def function(self, arg: str) -> None:
        print(f"Executing function with arg: {arg}")
    wrapped_function = wrapper(function)

obj = Class()
obj.function("hello")
obj.wrapped_function("world")
```
The errors from ty:
```
ty check /t/type.py                                                                                                   
error[missing-argument]: No argument provided for required parameter `arg`
  --> type.py:15:1
   |
15 | obj.wrapped_function("world")
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |

error[invalid-argument-type]: Argument is incorrect
  --> type.py:15:22
   |
15 | obj.wrapped_function("world")
   |                      ^^^^^^^ Expected `Class`, found `Literal["world"]`
   |

Found 2 diagnostics
```

### Version

ty version:
```sh
ty --version                                                                                                          
ty 0.0.57 (03cfaafe8 2026-07-08)
```

**Comment by carljm** [comment]
Thanks for the report! This is an unfortunate ambiguity in Python typing. Some callable objects (like functions) behave as method descriptors (that is, bind the first argument) when accessed on a class; some (like objects with `__call__` methods) do not. Both kinds of objects can inhabit a `Callable` type. `wrapper` is annotated to return a `Callable` type. In your particular example it actually returns a function, which is a method descriptor, but it could just as easily return some other kind of callable, which is not a method descriptor. There is no way to tell from the type annotations whether the returned callable is or isn't a method descriptor, so type checkers resort to heuristics. In this case ty's heuristic is a bit more conservative than mypy/pyright. If `wrapper` returned a callable object instead of a function, then ty's behavior would match runtime, and mypy would be wrong: 

```py
def wrapper[**P, R](method: Callable[P, R]) -> Callable[P, R]:
    class C:
        def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
            return method(*args, **kwargs)
    return C()
```

That said, I do think we should adjust our heuristic here, since `wrapper` is a transparent wrapper, to assume it returns a function. We already assume that if it's used via `@` decorator syntax, we should assume it with desugared call syntax too.

**Comment by kovidgoyal** [comment]
I am not sure I follow your reasoning here. Surely the type checker often knows what kind of object the wrapper is returning (from parsing its code) and it knows that the return value is being assigned to a class attribute. Those two pieces of information are enough to disambiguate the run time behavior, no? On the other hand I know very little about python typing so I could be missing something. 

Just for completeness, here is another form of this where ty complains about something that works at runtime and where the type annotation alone  actually has enough information for ty to disambiguate (I think):
```py
from collections.abc import Callable
from threading import Lock
from typing import Concatenate, Self

class Class:
    def wrapper[**P, R](method: Callable[Concatenate[Self, P], R]) -> Callable[Concatenate[Self, P], R]:
        def inner(self: Self, *args: P.args, **kwargs: P.kwargs) -> R:
            with self.lock:
                return method(self, *args, **kwargs)
        return inner

    lock: Lock = Lock()
    def function(self, arg: str) -> None:
        print(f"Executing function with arg: {arg}")
    wrapped_function = wrapper(function)

obj = Class()
obj.function("hello")
obj.wrapped_function("world")
```

**Comment by carljm** [comment]
Yes, in theory we could look inside the implementation of the function, even when it has a return annotation, and distinguish method-descriptor return types vs non-method-descriptor. I don't think any type checker does this (that I know of), but it is a more advanced approach we could implement that would give better results in some cases. (At some performance cost, particularly when the decorator is third-party: in that case we currently wouldn't check the implementation of the decorator at all.)

Your second example doesn't have enough information for ty to know for sure (the fact that a function takes `Self` as its first argument does not definitely make it a method descriptor at runtime), but it could be considered a strong heuristic signal.

The linked PR will fix both of these cases by just extending our existing heuristic for decorator syntax to the desugared version as well.

**Comment by kovidgoyal** [comment]
OK, I am fine with your PR as it handles my use case.
